client/state/data-layer/http-data/index.js

```diff
@@ -0,0 +1,66 @@
+/** @format */
+/**
+ * Internal dependencies
+ */
+import { HTTP_DATA_REQUEST, HTTP_DATA_TICK } from 'state/action-types';
+import { dispatchRequestEx } from 'state/data-layer/wpcom-http/utils';
+import { update } from './common';
+
+const fetch = action => {
+	update( action.id, 'pending' );
+
+	return [
+		{
+			...action.fetch,
+			onSuccess: action,
+			onFailure: action,
+			onProgress: action,
+		},
+		{ type: HTTP_DATA_TICK },
+	];
+};
+
+const onFailure = ( action, error ) => {
+	update( action.id, 'failure', error );
+
+	return { type: HTTP_DATA_TICK };
+};
+
+const parseResponse = ( data, fromApi ) => {
+	if ( 'function' !== typeof fromApi ) {
+		return data;
+	}
+
+	// [ error, data ]
+	try {
+		return [ undefined, fromApi( data ) ];
+	} catch ( error ) {
+		return [ error, undefined ];
+	}
+};
+
+const onSuccess = ( action, apiData ) => {
+	const [ error, data ] = parseResponse( apiData, action.fromApi );
```

Comment:
whats the reason for the `parseResponse` function as opposed to lifting the try/catch to this function? From what I can tell, most of parseResponse is converting a try/catch into an if/else.

we could also write:

```js
const onSuccess = (action, apiData) => {
  let data;
  try {
     data = action.fromApi(apiData);
  } catch( error ) {
     return onFailure(action, error) 
  }
  // rest of func
}
```

Reply:
only that ugly `let` :wink: